RepositoryCompiler/CodeModel/CodeParsers/CSharp/CaDETMemberMetricCalculator.cs

```diff
@@ -121,5 +137,404 @@ private int CountOccurrences(string text, string pattern)
+        }
+
+        private int CountNumberOfNumericLiterals(MemberDeclarationSyntax method)
+        {
+            return method.DescendantNodes().OfType<LiteralExpressionSyntax>().Count(n => n.IsKind(SyntaxKind.NumericLiteralExpression));
+        }
+
+        private int CountNumberOfStringLiterals(MemberDeclarationSyntax method)
+        {
+            return method.DescendantNodes().OfType<LiteralExpressionSyntax>().Count(n => n.IsKind(SyntaxKind.StringLiteralExpression));
+        }
+
+        private int CountNumberOfMathOperations(MemberDeclarationSyntax method)
+        {
+            int count = CountBinaryExpressions(method);
+            count += CountUnaryExpressions(method);
+            return count;
+        }
+
+        private int CountBinaryExpressions(MemberDeclarationSyntax method)
+        {
+            return method.DescendantNodes().OfType<BinaryExpressionSyntax>()
+                .Count(n => n.IsKind(SyntaxKind.AddExpression) ||
+                            n.IsKind(SyntaxKind.SubtractExpression) ||
+                            n.IsKind(SyntaxKind.MultiplyExpression) ||
+                            n.IsKind(SyntaxKind.DivideExpression) ||
+                            n.IsKind(SyntaxKind.ModuloExpression) ||
+                            n.IsKind(SyntaxKind.LeftShiftExpression) ||
+                            n.IsKind(SyntaxKind.RightShiftExpression));
+        }
+
+        private int CountUnaryExpressions(MemberDeclarationSyntax method)
+        {
+            int count = method.DescendantNodes().OfType<PrefixUnaryExpressionSyntax>()
+                .Count(n => n.IsKind(SyntaxKind.PreIncrementExpression) ||
+                            n.IsKind(SyntaxKind.PreDecrementExpression) ||
+                            n.IsKind(SyntaxKind.UnaryPlusExpression) ||
+                            n.IsKind(SyntaxKind.UnaryMinusExpression));
+            count += method.DescendantNodes().OfType<PostfixUnaryExpressionSyntax>()
+                .Count(n => n.IsKind(SyntaxKind.PostIncrementExpression) ||
+                            n.IsKind(SyntaxKind.PostDecrementExpression));
+            return count;
+        }
+
+        private int CountNumberOfParenthesizedExpressions(MemberDeclarationSyntax method)
+        {
+            return method.DescendantNodes().OfType<ExpressionSyntax>()
+                .Count(n => n.IsKind(SyntaxKind.ParenthesizedExpression));
+        }
+
+        private int CountNumberOfLambdaExpressions(MemberDeclarationSyntax method)
+        {
+            return method.DescendantNodes().OfType<LambdaExpressionSyntax>().Count();
+        }
+
+        private int CountMaxNestedBlocks(MemberDeclarationSyntax method)
+        {
+            if (!method.Kind().Equals(SyntaxKind.MethodDeclaration))
+            {
+                return 0;
+            }
+            BaseMethodDeclarationSyntax baseMethod = (BaseMethodDeclarationSyntax)method;
```

Comment:
Why is this cast important? Can we not extract the descendant nodes from the member, or will this include ancestors from the class and namespace level?

Reply:
I was thinking of avoiding the block that represents the body of the method by extracting descendant nodes from the body (and in order to access the body, I had to cast). Now I see that it is not really the problem (the number of nested blocks is calculated correctly without it), so I will remove it.